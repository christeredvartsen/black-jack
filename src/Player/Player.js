import React from 'react';
import PropTypes from 'prop-types';
import types from '../propTypes';
import './Player.scss';

const cardKey = (c) => c.value + c.suit.charAt(0);

const Player = ({ drawCard, name, score, cards, draw }) => {
  return (
    <fieldset className="player">
      <legend>{name}</legend>
      <dl>
        <dt>Score</dt>
        <dd className="playerScore">{score}</dd>
        <dt>Cards</dt>
        <dd className="playerCards">
          {cards.length > 0 ? (
            <ol>{cards.map((card) => <li key={cardKey(card)}>{cardKey(card)}</li>)}</ol>
          ) : <p>The player has no cards.</p>}
        </dd>
      </dl>
      {draw ? <button onClick={drawCard}>Draw</button> : null}
    </fieldset>
  );
};
Player.propTypes = {
  drawCard: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  score: PropTypes.number.isRequired,
  cards: types.cards.isRequired,
  draw: PropTypes.bool.isRequired,
};

export default Player;