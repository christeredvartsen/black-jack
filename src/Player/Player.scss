fieldset.player {
  margin-bottom: 1em;
  flex-grow: 1;

  legend {
    font-size: 1.5em;
  }
  
  dl dd ol {
    list-style-type: none;
    margin-top: 0;
    padding-left: 0;
  }
}