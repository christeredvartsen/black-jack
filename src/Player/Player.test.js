import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Player from './Player';

describe('<Player />', () => {
  const cards = [
    { suit: 'HEARTS', value: '10' },
    { suit: 'HEARTS', value: 'A' },
    { suit: 'SPADES', value: 'A' },
  ];

  test('does not render draw button for all players', () => {
    expect(
      shallow(
        <Player cards={cards} draw={false} drawCard={jest.fn()} name="some name" score={0} />
      ).find('button').exists()
    ).toBe(false);
  });

  test('can draw cards', () => {
    const draw = jest.fn();
    const component = shallow(
      <Player cards={cards} draw={true} drawCard={draw} name="some name" score={0} />
    );
    component.find('button').simulate('click');
    expect(draw).toHaveBeenCalledTimes(1);
  });

  test('generates unique keys for cards', () => {
    const component = shallow(
      <Player cards={cards} draw={true} drawCard={jest.fn()} name="some name" score={32} />
    );
    ['10H', 'AH', 'AS'].forEach(
      (expectedKey, i) => expect(
        component.find('ol li').at(i).key()
      ).toBe(expectedKey)
    );
  });

  test('matches snapshot', () => {
    const tree = shallow(
      <Player draw={true} drawCard={jest.fn()} name="Some name" score={0} cards={cards} />
    );
    expect(toJson(tree)).toMatchSnapshot();
  });
});