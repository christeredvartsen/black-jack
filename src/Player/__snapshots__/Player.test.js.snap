// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Player /> matches snapshot 1`] = `
<fieldset
  className="player"
>
  <legend>
    Some name
  </legend>
  <dl>
    <dt>
      Score
    </dt>
    <dd
      className="playerScore"
    >
      0
    </dd>
    <dt>
      Cards
    </dt>
    <dd
      className="playerCards"
    >
      <ol>
        <li
          key="10H"
        >
          10H
        </li>
        <li
          key="AH"
        >
          AH
        </li>
        <li
          key="AS"
        >
          AS
        </li>
      </ol>
    </dd>
  </dl>
  <button
    onClick={[MockFunction]}
  >
    Draw
  </button>
</fieldset>
`;
